lib/user.js

```diff
@@ -0,0 +1,111 @@
+var keystone = require('../'),
+	crypto = require('crypto'),
+	async = require('async'),
+	moment = require('moment');
+
+exports.init = function() {
+	var mongoose = keystone.mongoose,
+		Reset = new mongoose.Schema({
+			token  : {type: String, index: true},
+			expires: {type: Date, default: Date.now}
+		}, {collection: keystone.prefixModel('App_User')});
+	mongoose.model('App_User', Reset);
+};
+
+/**
+ * Checks if a user with the given the lookup email exists in the database
+ *
+ * @param {Object} lookup - must contain email
+ * @param {function()} onSuccess callback, is passed the User instance
+ * @param {function()} onFail callback
+ */
+
+exports.attemptReset = function(lookup, next) {
+	var User = keystone.list(keystone.get('user model'));
+	if ('string' === typeof lookup.email) {
+		// match email address
+		User.model.findOne({email: lookup.email}).exec(function(err, user) {
+			if (err) {
+				return next(err);
+			}
+			if (user && user.sendResetPassword) {
+				return user.sendResetPassword(next);
+			}
+			next();
+		});
+	} else {
+		next(new Error('todo: e-mail must be string'));
+	}
+};
+
+exports.attemptSignin = function(token, req, res, next) {
+	async.waterfall([
+		function(callback) {
+			var mongoose = keystone.mongoose,
+				Reset = mongoose.model('App_User');
+			Reset.findOne({token: token}).exec(function(err, reset){
+				callback(err, reset);
+			});
+		},
+		function(reset, callback){
+			var userId,
+				err;
+			if(reset){
+				if(!reset.expires || moment().diff(reset.expires) >= 0 || !reset.token || reset.token !== token){
```

Comment:
Can you simplify this line instead to be:

``` javascript
if (!reset.expires || !reset.token) return callback (new TypeError("Invalid reset data"))
if (moment().diff(reset.expires) >= 0)) return callback(new Error("Reset token expired"))
if (reset.token !== token) return callback(new Error("Reset token does not match"))
```

There is also a timing attack possibility in this last line, not sure if that needs to be mitigated.